[HDR] Scan the layer backing for HDR contents

shallawa opened the pull request:
#### 0acfecd15ef377be44bb88abd4649cb973a0075f
<pre>
[HDR] Scan the layer backing for HDR contents
<a href="https://bugs.webkit.org/show_bug.cgi?id=287411">https://bugs.webkit.org/show_bug.cgi?id=287411</a>
<a href="https://rdar.apple.com/144534518">rdar://144534518</a>

Reviewed by Simon Fraser.

This work is towards creating HDR layers only when needed. See bug 282298.

When the render sub-tree of a layer is scanned for painted contents, we can check
for the HDR contents also.

PaintedContentRequest will have a new RequestState named `hasPaintedHDRContent`.
PaintedContentsInfo will have a new RequestState named `m_hdrContent`.

PaintedContentsInfo::determinePaintsContent() will determine the paintedContent
and the paintsHDRContent in one traversal by calling
RenderLayerBacking::determinePaintsContent(). When this function returns,
`hasPaintedHDRContent` will be copied to `m_hdrContent`.

determineNonLayerDescendantsPaintedContent() will be changed that such it will
traverse the full render tree until PaintedContentRequest::hasPaintedHDRContent
is satisfied. To optimize the cases where the feature flags are disabled, the
screen is not HDR or the caller does not want to know whether there is HDR content
or not, PaintedContentRequest::hasPaintedHDRContent will be set to DontCare in
these cases.

PaintedContentRequest::isSatisfied() will return true if both hasPaintedContent
and hasPaintedHDRContent are satisfied.

In future patches, a new method RenderLayerBacking::updateDrawsHDRContent() will
be added. It will force calling PaintedContentsInfo::determinePaintsContent() if
it was not called before and it will update the ContentsFormat of its
`m_graphicsLayer`.

* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::drawsHDRContent const):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::pixelFormat const):
Make it easy to answer the question: Does this canvas element requrie HDR layer?

* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::headroom const):
* Source/WebCore/loader/cache/CachedImage.h:
Make it easy to answer the question: Does this image element requrie HDR layer?

* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/BitmapImageDescriptor.cpp:
(WebCore::BitmapImageDescriptor::headroom const):
* Source/WebCore/platform/graphics/BitmapImageDescriptor.h:
* Source/WebCore/platform/graphics/BitmapImageSource.cpp:
(WebCore::BitmapImageSource::frameHeadroomAtIndex const): Deleted.
* Source/WebCore/platform/graphics/BitmapImageSource.h:
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::headroom const):
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::frameOrientationAtIndex const):
(WebCore::ImageSource::frameHeadroomAtIndex const): Deleted.
* Source/WebCore/platform/graphics/NativeImageSource.h:
Make the headroom a metadata of the Image. Currently the headroom is a metadata
of NativeImage. This requires decoding the Image before painting it. So it will
eliminate the async image decoding entirely. Ideally, the headroom should be a
metadata of the ImageDecoder. Like ImageDecoder::frameSizeAtIndex(), we should
have something like ImageDecoder::frameHeadroomAtIndex().

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::PaintedContentRequest::PaintedContentRequest):
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::PaintedContentsInfo::paintsBoxDecorations):
(WebCore::PaintedContentsInfo::paintsContent):
(WebCore::PaintedContentsInfo::paintsHDRContent):
(WebCore::PaintedContentsInfo::determinePaintsBoxDecorations):
(WebCore::PaintedContentsInfo::determinePaintsContent):
(WebCore::RenderLayerBacking::determinePaintsContent const):
(WebCore::RenderLayerBacking::determineNonCompositedLayerDescendantsPaintedContent const):
(WebCore::PaintedContentsInfo::paintsBoxDecorationsDetermination): Deleted.
(WebCore::PaintedContentsInfo::paintsContentDetermination): Deleted.
(WebCore::RenderLayerBacking::paintsContent const): Deleted.
(WebCore::RenderLayerBacking::isPaintDestinationForDescendantLayers const): Deleted.
* Source/WebCore/rendering/RenderLayerBacking.h:

Canonical link: <a href="https://commits.webkit.org/290274@main">https://commits.webkit.org/290274@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d13d02d2bbb7ea4933527fac83eac68d0d970cd1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89519 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9046 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44376 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94511 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40286 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91571 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9435 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17325 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68955 "Found 1 new test failure: media/media-fullscreen-inline.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26610 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92520 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7239 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81239 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49321 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6970 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/35619 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39392 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77324 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36624 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96339 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16701 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77827 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16956 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77040 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77143 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19031 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21570 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20148 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9861 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16714 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22027 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16455 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19906 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18237 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -5768,6 +5768,34 @@ static const unsigned maxRendererTraversalCount = 200;
+#endif
+
     for (const auto& child : childrenOfType<RenderObject>(renderer)) {
         if (++renderersTraversed > maxRendererTraversalCount) {
```
We need to not early return here if the caller needs to know the "hasPaintedHDRContent" status (which we need to indicate via `RequestState`).

smfr commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -561,25 +561,34 @@ class RenderLayer final : public CanMakeSingleThreadWeakPtr<RenderLayer>, public
 
-        bool probablyHasPaintedContent() const { return hasPaintedContent == RequestState::True || hasPaintedContent == RequestState::Undetermined; }
-        
         bool isSatisfied() const { return hasPaintedContent != RequestState::Unknown; }
```
You need to change this too. The request is only satisfied if all the questions have been answered (i.e. the caller wants to know about both `hasPaintedContent` and `hasPaintedHDRContent` and we know the state for both).

shallawa replied:
You are right. I forgot to port the change in this function from https://github.com/WebKit/WebKit/pull/40138.

smfr commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -557,29 +557,39 @@ class RenderLayer final : public CanMakeSingleThreadWeakPtr<RenderLayer>, public
-        bool probablyHasPaintedContent() const { return hasPaintedContent == RequestState::True || hasPaintedContent == RequestState::Undetermined; }
-        
-        bool isSatisfied() const { return hasPaintedContent != RequestState::Unknown; }
+        bool isSatisfied() const
```
This still needs to have an "or".

I think the fact that we're struggling to get this right makes it clear that was need to be able to write tests for this.

shallawa committed:
[HDR] Scan the layer backing for HDR contents
https://bugs.webkit.org/show_bug.cgi?id=287411
rdar://144534518

Reviewed by Simon Fraser.

This work is towards creating HDR layers only when needed. See bug 282298.

When the render sub-tree of a layer is scanned for painted contents, we can check
for the HDR contents also.

PaintedContentRequest will have a new RequestState named `hasPaintedHDRContent`.
PaintedContentsInfo will have a new RequestState named `m_hdrContent`.

PaintedContentsInfo::determinePaintsContent() will determine the paintedContent
and the paintsHDRContent in one traversal by calling
RenderLayerBacking::determinePaintsContent(). When this function returns,
`hasPaintedHDRContent` will be copied to `m_hdrContent`.

determineNonLayerDescendantsPaintedContent() will be changed that such it will
traverse the full render tree until PaintedContentRequest::hasPaintedHDRContent
is satisfied. To optimize the cases where the feature flags are disabled, the
screen is not HDR or the caller does not want to know whether there is HDR content
or not, PaintedContentRequest::hasPaintedHDRContent will be set to DontCare in
these cases.

PaintedContentRequest::isSatisfied() will return true if both hasPaintedContent
and hasPaintedHDRContent are satisfied.

In future patches, a new method RenderLayerBacking::updateDrawsHDRContent() will
be added. It will force calling PaintedContentsInfo::determinePaintsContent() if
it was not called before and it will update the ContentsFormat of its
`m_graphicsLayer`.

* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::drawsHDRContent const):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::pixelFormat const):
Make it easy to answer the question: Does this canvas element requrie HDR layer?

* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::headroom const):
* Source/WebCore/loader/cache/CachedImage.h:
Make it easy to answer the question: Does this image element requrie HDR layer?

* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/BitmapImageDescriptor.cpp:
(WebCore::BitmapImageDescriptor::headroom const):
* Source/WebCore/platform/graphics/BitmapImageDescriptor.h:
* Source/WebCore/platform/graphics/BitmapImageSource.cpp:
(WebCore::BitmapImageSource::frameHeadroomAtIndex const): Deleted.
* Source/WebCore/platform/graphics/BitmapImageSource.h:
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::headroom const):
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::frameOrientationAtIndex const):
(WebCore::ImageSource::frameHeadroomAtIndex const): Deleted.
* Source/WebCore/platform/graphics/NativeImageSource.h:
Make the headroom a metadata of the Image. Currently the headroom is a metadata
of NativeImage. This requires decoding the Image before painting it. So it will
eliminate the async image decoding entirely. Ideally, the headroom should be a
metadata of the ImageDecoder. Like ImageDecoder::frameSizeAtIndex(), we should
have something like ImageDecoder::frameHeadroomAtIndex().

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::PaintedContentRequest::PaintedContentRequest):
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::PaintedContentsInfo::paintsBoxDecorations):
(WebCore::PaintedContentsInfo::paintsContent):
(WebCore::PaintedContentsInfo::paintsHDRContent):
(WebCore::PaintedContentsInfo::determinePaintsBoxDecorations):
(WebCore::PaintedContentsInfo::determinePaintsContent):
(WebCore::RenderLayerBacking::determinePaintsContent const):
(WebCore::RenderLayerBacking::determineNonCompositedLayerDescendantsPaintedContent const):
(WebCore::PaintedContentsInfo::paintsBoxDecorationsDetermination): Deleted.
(WebCore::PaintedContentsInfo::paintsContentDetermination): Deleted.
(WebCore::RenderLayerBacking::paintsContent const): Deleted.
(WebCore::RenderLayerBacking::isPaintDestinationForDescendantLayers const): Deleted.
* Source/WebCore/rendering/RenderLayerBacking.h:

Canonical link: https://commits.webkit.org/290274@main
